src/transactions/TransactionSQL.cpp

```diff
@@ -8,14 +8,338 @@
+                  uint32_t ledgerCount)
+{
+    ZoneScoped;
+    auto& db = app.getDatabase();
+    auto timer = db.getSelectTimer("txsethistory");
+    std::string base64TxSet;
+    uint32_t begin = ledgerSeq, end = ledgerSeq + ledgerCount;
+    uint32_t curLedgerSeq;
+    releaseAssert(begin <= end);
+    soci::statement st =
+        (sess.prepare << "SELECT ledgerseq, txset FROM txsethistory "
+                         "WHERE ledgerseq >= :begin AND ledgerseq < :end ORDER "
+                         "BY ledgerseq ASC",
+         soci::into(curLedgerSeq), soci::into(base64TxSet), soci::use(begin),
+         soci::use(end));
+
+    std::vector<std::pair<uint32_t, std::vector<uint8_t>>> txSets;
+
+    st.execute(true);
+
+    while (st.got_data())
+    {
+        auto& [txSetLedgerSeq, encodedTxSet] = txSets.emplace_back();
+        txSetLedgerSeq = curLedgerSeq;
+        decoder::decode_b64(base64TxSet, encodedTxSet);
+        st.fetch();
+    }
+    return txSets;
+}
+
+void
+writeNonGeneralizedTxSetToStream(
+    Database& db, soci::session& sess, uint32 ledgerSeq,
+    std::vector<TransactionFrameBasePtr> const& txs,
+    TransactionHistoryResultEntry& results, XDROutputFileStream& txOut,
+    XDROutputFileStream& txResultOut)
+{
+    ZoneScoped;
+    // prepare the txset for saving
+    auto lh = LedgerHeaderUtils::loadBySequence(db, sess, ledgerSeq);
+    if (!lh)
+    {
+        throw std::runtime_error("Could not find ledger");
+    }
+    auto txSet =
+        TxSetFrame::makeFromHistoryTransactions(lh->previousLedgerHash, txs);
+    TransactionHistoryEntry hist;
+    hist.ledgerSeq = ledgerSeq;
+    txSet->toXDR(hist.txSet);
+    txOut.writeOne(hist);
+    txResultOut.writeOne(results);
+}
+
+void
+writeGeneralizedTxSetToStream(Database& db, soci::session& sess,
+                              uint32 ledgerSeq,
+                              std::vector<uint8_t> const& encodedTxSet,
+                              std::vector<TransactionFrameBasePtr> const& txs,
+                              TransactionHistoryResultEntry& results,
+                              XDROutputFileStream& txOut,
+                              XDROutputFileStream& txResultOut)
+{
+    ZoneScoped;
+    UnorderedMap<Hash, TransactionFrameBase const*> txByHash;
+    for (auto const& tx : txs)
+    {
+        txByHash[tx->getFullHash()] = tx.get();
+    }
+
+    GeneralizedTxSetUnpacker unpacker(encodedTxSet, txByHash);
+    TransactionHistoryEntry hist;
+    hist.ledgerSeq = ledgerSeq;
+    hist.ext.v(1);
+    xdr_argpack_archive(unpacker, hist.ext.generalizedTxSet());
+
+    txOut.writeOne(hist);
+    txResultOut.writeOne(results);
+}
+
+// The following method handles the protocol version upgrade. Before
+// GENERALIZED_TX_SET_PROTOCOL_VERSION we don't store the tx set at
+// all, hence the results are empty. Starting from
+// GENERALIZED_TX_SET_PROTOCOL_VERSION we do store it. However, in
+// order to figure out if ledger `N` has generalized tx set we need
```

Comment:
I see what you mean. That's not correct though: in published ledger headers, upgrade ledger contains the hash of an old-style tx set, so published tx sets must match what's in ledger headers, in order for core to properly verify history (otherwise the replay will fork). We do the check here: https://github.com/stellar/stellar-core/blob/ce981331000128a1c145ec364fbd83d3dd4be5ed/src/catchup/ApplyCheckpointWork.cpp#L219 So I think we should keep the tx set at the upgrade ledger as-is.